drv/fpga-devices/src/ecp5.rs

```diff
@@ -0,0 +1,453 @@
+    ///
+    /// See FPGA-TN-02039-2.0, 4.5.3 for details.
+    fn init_n(&self) -> Result<bool, Self::Error>;
+    fn set_init_n(&self, asserted: bool) -> Result<(), Self::Error>;
+
+    /// DONE interface. This pin signals the device is in User Mode. Asserting
+    /// the pin keeps the device from entering User Mode after Configuration.
+    ///
+    /// See FPGA-TN-02039-2.0, 4.5.4 for details.
+    fn done(&self) -> Result<bool, Self::Error>;
+    fn set_done(&self, asserted: bool) -> Result<(), Self::Error>;
+
+    /// Application Reset.
+    fn application_reset_n(&self) -> Result<bool, Self::Error>;
+    fn set_application_reset_n(
+        &self,
+        asserted: bool,
+    ) -> Result<(), Self::Error>;
+
+    /// A generic interface to send commands and read/write data from a
+    /// configuration port. This interface is intended to be somewhat transport
+    /// agnostic so either SPI or JTAG could be implemented if desired.
+    fn write_command(&self, c: Command) -> Result<(), Self::Error>;
+    fn read(&self, buf: &mut [u8]) -> Result<(), Self::Error>;
+    fn write(&self, buf: &[u8]) -> Result<(), Self::Error>;
+
+    /// The command interface may exist on a shared medium such as SPI. The
+    /// following primitives allow the upper half of the driver to issue atomic
+    /// commands.
+    ///
+    /// If no lock control of the medium is needed these can be implemented as
+    /// no-op.
+    fn lock(&self) -> Result<(), Self::Error>;
+    fn release(&self) -> Result<(), Self::Error>;
+}
+
+/// Newtype wrapping an Impl, allowing the remaining traits to be implemented.
+pub struct Ecp5<ImplT: Ecp5Impl>(ImplT);
+
+/// Additional ECP5 methods to help implement high level behavior.
+impl<ImplT: Ecp5Impl> Ecp5<ImplT> {
+    /// Send a command to the device which does not return or require additional
+    /// data. FPGA-TN-02039-2.0, 6.2.5 refers to this as a Class C command.
+    pub fn send_command(&self, c: Command) -> Result<(), ImplT::Error> {
+        self.0.lock()?;
+        self.0.write_command(c)?;
+        self.0.release()?;
+        ringbuf_entry!(Trace::Command(c));
+        Ok(())
+    }
+
+    /// Send a command and read back a number of bytes given by the type T. Note
+    /// that data is always returned in big endian order.
+    pub fn read<T: Default + AsBytes + FromBytes>(
+        &self,
+        c: Command,
+    ) -> Result<T, ImplT::Error> {
+        let mut buf = T::default();
+
+        self.0.lock()?;
+        self.0.write_command(c)?;
+        self.0.read(buf.as_bytes_mut())?;
+        self.0.release()?;
+
+        Ok(buf)
+    }
+
+    /// Send a `Command` and read back two bytes of data.
+    pub fn read16(&self, c: Command) -> Result<u16, ImplT::Error> {
+        Ok(u16::from_be(self.read(c)?))
+    }
+
+    /// Send a `Command` and read back four bytes of data.
+    pub fn read32(&self, c: Command) -> Result<u32, ImplT::Error> {
+        let v = u32::from_be(self.read(c)?);
+
+        match c {
```

Comment:
Nit: I'd rather have this `ringbuf` logging in `status` and `device_id`, since `read32`'s name implies that it's low-level and agnostic to the value of `c`

Reply:
This was removed.